ftms_filtering.qmd

```diff
@@ -119,179 +126,195 @@ perc_ms2_assigned <- (n_assigned_ms2 / n_total_ms2) * 100
-Thus, each MS2 spectrum is finally linked to only one feature, while alternative
-links are marked as ambiguous.
+Note that an MS2 spectrum could also be assigned to more than one feature. The
+total number of MS2 spectra assigned to more than one feature is:
+
+```{r}
+ms2_id <- paste0(ms2$dataOrigin, ms2$acquisitionNum)
+ms2_mult <- names(table(ms2_id)[table(ms2_id) > 1])
+length(ms2_mult)
+```
+
+Manually evaluate some of these cases:
+
+```{r}
+a <- ms2[ms2_id == ms2_mult[3L]]
+
+a_chr <- featureChromatograms(ftms, features = a$feature_id)
+par(mfrow = c(1, 2))
+plot(a_chr[1])
+abline(v = a$rtime, col = "red")
+plot(a_chr[2])
+abline(v = a$rtime, col = "red")
+```
+
+The features are thus very close in the retention time dimension, eventually
+even not well separated. The MS2 spectrum (retention time highlighted in red) is
+associated to a chromatographic peak in one sample that was eventually wrongly
+assigned to the earlier feature.
+
+For most other examples, the feature seem to represent noise or at least no
+clear chromatographic signal (see for example below).
+
+```{r}
+a <- ms2[ms2_id == ms2_mult[200L]]
+
+a_chr <- featureChromatograms(ftms, features = a$feature_id)
+par(mfrow = c(1, 2))
+plot(a_chr[1])
+abline(v = a$rtime, col = "red")
+plot(a_chr[2])
+abline(v = a$rtime, col = "red")
+```
+
+**Question**: (maybe for the next meeting with Rebecca) should we maybe consider
+keeping only *high quality* signals?  trying to remove these noisy
+features/chromatographic peaks?
+
+
+## Resolving multiple assignments of MS2 spectra to features
+
+We next evaluate the cases above to reduce/remove the multiple assignments and
+keep only a single MS2-feature association.
+First, we select the duplicated MS2 spectra by
+identifying those with the same *dataOrigin* and *acquisitionNum* but different
+*feature_id* to detect duplicates assigned to multiple features. From these
+duplicated MS2 spectra, we then check first (*case A*) if a feature has unique
+MS2 spectra elsewhere. If that is the case any ambiguous assignment is removed
+and only the unambiguous mappings are kept for the feature. Subsequently, the
+assignments are checked again and if there are still duplicated assignments the
+distance of the MS2 to the respective features is evaluated (*case B*): we keep
+the top 10% MS2-feature assignments. All other assignments are dropped for the
+respective feature.
 
 ```{r}
 select_unique_MS2 <- function(ms2, ftms) {
-  
+
   sp <- spectraData(ms2)
-  
+
   # Create a unique identifier for each unique groupe of (dataOrigin, scanIndex)
   sp$groups <- paste0(sp$dataOrigin, "_", sp$scanIndex)
-  
-  # Create indices for duplicated groups 
+
+  # Create indices for duplicated groups
   group_idx <- split(seq_len(nrow(sp)), sp$groups)
-  
+
+  # define MS2 as ambiguous if they are duplicated in sp (i.e., are assigned
+  # to >= 2 features. We use the unique ID for the MS2 spectra and their
+  # indices in sp for that.
+  sp$is_ambiguous <- FALSE
```

Comment:
I added this new column which is `TRUE` for all MS2 spectra that are assigned to more than one feature and `FALSE` otherwise.